File: cpp/src/elements/FrameBufferView.cpp
#include "ui/elements/FrameBufferView.hpp"
#include "ui/core/TextVisual.hpp"

#include <algorithm>
#include <cmath>
#include <iomanip>
#include <sstream>
#include <tuple>
#include <utility>

namespace ui {

namespace {

[[nodiscard]] std::optional<Color> colorFromString(const std::string& value) {
    if (value.empty()) {
        return std::nullopt;
    }
    try {
        return Color(value);
    } catch (...) {
        return std::nullopt;
    }
}

[[nodiscard]] std::string dimHexColor(const std::string& value, double dimAlpha = 0.4) {
    if (value.size() != 7 || value.front() != '#') {
        return value;
    }
    try {
        const int red = static_cast<int>(std::round((1.0 - dimAlpha) * std::stoi(value.substr(1, 2), nullptr, 16)));
        const int green = static_cast<int>(std::round((1.0 - dimAlpha) * std::stoi(value.substr(3, 2), nullptr, 16)));
        const int blue = static_cast<int>(std::round((1.0 - dimAlpha) * std::stoi(value.substr(5, 2), nullptr, 16)));
        std::ostringstream out;
        out << "#" << std::hex << std::setfill('0') << std::setw(2) << red
            << std::setw(2) << green << std::setw(2) << blue;
        return out.str();
    } catch (...) {
        return value;
    }
}

[[nodiscard]] std::vector<int> splitDimension(int size, int parts) {
    size = std::max(1, size);
    parts = std::max(1, std::min(parts, size));
    const int base = size / parts;
    const int extra = size % parts;
    std::vector<int> result;
    result.reserve(static_cast<std::size_t>(parts));
    for (int index = 0; index < parts; ++index) {
        result.push_back(base + (index < extra ? 1 : 0));
    }
    return result;
}

[[nodiscard]] std::string firstCellGlyph(const std::string& text) {
    return safeTerminalCellText(text);
}

}  // namespace

FrameBufferView::FrameBufferView(std::string name) : Element(std::move(name)) {}

void FrameBufferView::clearFrames() {
    frames_.clear();
    targetOrder_.clear();
    regions_.clear();
}

void FrameBufferView::setFrames(std::map<std::string, std::vector<std::vector<FrameBufferCell>>> frames,
                                std::vector<std::string> targetOrder,
                                std::map<std::string, FrameBufferRegion> regions) {
    frames_ = std::move(frames);
    targetOrder_ = std::move(targetOrder);
    regions_ = std::move(regions);
}

void FrameBufferView::setPan(int row, int col) {
    panRow_ = std::max(0, row);
    panCol_ = std::max(0, col);
}

void FrameBufferView::setDimmed(bool dimmed) {
    dimmed_ = dimmed;
}

void FrameBufferView::setInteractionState(bool focused, bool edit) {
    panelFocused_ = focused;
    panelEdit_ = edit;
}

std::map<std::string, FrameBufferRegion> FrameBufferView::defaultRegions(Size size) const {
    std::map<std::string, FrameBufferRegion> result;
    const int count = std::max(1, static_cast<int>(targetOrder_.size()));
    const std::vector<int> widths = splitDimension(std::max(1, size.width), count);
    int col = 0;
    for (int index = 0; index < static_cast<int>(targetOrder_.size()); ++index) {
        result[targetOrder_[static_cast<std::size_t>(index)]] = FrameBufferRegion{
            .row = 0,
            .col = col,
            .width = widths[static_cast<std::size_t>(index)],
            .height = std::max(1, size.height),
        };
        col += widths[static_cast<std::size_t>(index)];
    }
    return result;
}

void FrameBufferView::overlayCorners(RenderedContent& content,
                                     const std::map<std::string, FrameBufferRegion>& regions,
                                     const Style& style) const {
    if (!panelFocused_ && !panelEdit_) {
        return;
    }
    const Style& stateStyle = panelEdit_ && editStyle().has_value()
        ? *editStyle()
        : (panelFocused_ && focusStyle().has_value() ? *focusStyle() : style);
    const std::optional<Color> color = stateStyle.color.has_value()
        ? stateStyle.color
        : std::optional<Color>(Color(panelEdit_ ? "#22c55e" : "#facc15"));
    for (const std::string& name : targetOrder_) {
        auto regionIt = regions.find(name);
        if (regionIt == regions.end()) {
            continue;
        }
        const FrameBufferRegion& region = regionIt->second;
        if (region.width < 2 || region.height < 2) {
            continue;
        }
        const std::vector<std::tuple<int, int, std::string>> points = {
            {region.row, region.col, "┌"},
            {region.row, region.col + 1, "─"},
            {region.row, region.col + region.width - 2, "─"},
            {region.row, region.col + region.width - 1, "┐"},
            {region.row + region.height - 1, region.col, "└"},
            {region.row + region.height - 1, region.col + 1, "─"},
            {region.row + region.height - 1, region.col + region.width - 2, "─"},
            {region.row + region.height - 1, region.col + region.width - 1, "┘"},
        };
        for (const auto& [row, col, text] : points) {
            if (row >= 0 && row < static_cast<int>(content.size()) &&
                col >= 0 && col < static_cast<int>(content[static_cast<std::size_t>(row)].size())) {
                TerminalCell& cell = content[static_cast<std::size_t>(row)][static_cast<std::size_t>(col)];
                cell.text = text;
                cell.foreground = color;
            }
        }
    }
}

RenderedContent FrameBufferView::render(Size size, ElementRenderState state) const {
    const Style style = effectiveStyle(state.focused, state.editMode);
    const int width = std::max(1, size.width);
    const int height = std::max(1, size.height);
    RenderedContent content;
    content.reserve(static_cast<std::size_t>(height));
    for (int row = 0; row < height; ++row) {
        RenderedRow renderedRow;
        renderedRow.reserve(static_cast<std::size_t>(width));
        for (int col = 0; col < width; ++col) {
            renderedRow.push_back(TerminalCell{.text = " ", .foreground = style.color, .background = style.background});
        }
        content.push_back(std::move(renderedRow));
    }

    const std::map<std::string, FrameBufferRegion> regions = regions_.empty() ? defaultRegions(size) : regions_;
    for (const std::string& name : targetOrder_) {
        auto regionIt = regions.find(name);
        if (regionIt == regions.end()) {
            continue;
        }
        auto frameIt = frames_.find(name);
        if (frameIt == frames_.end()) {
            continue;
        }
        const FrameBufferRegion& region = regionIt->second;
        const auto& rows = frameIt->second;
        for (int localRow = 0; localRow < region.height; ++localRow) {
            const int outRow = region.row + localRow;
            if (outRow < 0 || outRow >= height) {
                continue;
            }
            const int sourceRow = panRow_ + localRow;
            if (sourceRow < 0 || sourceRow >= static_cast<int>(rows.size())) {
                continue;
            }
            const auto& sourceCells = rows[static_cast<std::size_t>(sourceRow)];
            for (int localCol = 0; localCol < region.width; ++localCol) {
                const int outCol = region.col + localCol;
                if (outCol < 0 || outCol >= width) {
                    continue;
                }
                const int sourceCol = panCol_ + localCol;
                if (sourceCol < 0 || sourceCol >= static_cast<int>(sourceCells.size())) {
                    continue;
                }
                FrameBufferCell cell = sourceCells[static_cast<std::size_t>(sourceCol)];
                if (dimmed_) {
                    cell.foreground = dimHexColor(cell.foreground);
                    cell.background = dimHexColor(cell.background);
                }
                content[static_cast<std::size_t>(outRow)][static_cast<std::size_t>(outCol)] = TerminalCell{
                    .text = firstCellGlyph(cell.text),
                    .foreground = colorFromString(cell.foreground).value_or(style.color.value_or(Color())),
                    .background = colorFromString(cell.background).value_or(style.background.value_or(Color())),
                };
                if (!colorFromString(cell.foreground).has_value() && !style.color.has_value()) {
                    content[static_cast<std::size_t>(outRow)][static_cast<std::size_t>(outCol)].foreground = std::nullopt;
                }
                if (!colorFromString(cell.background).has_value() && !style.background.has_value()) {
                    content[static_cast<std::size_t>(outRow)][static_cast<std::size_t>(outCol)].background = std::nullopt;
                }
            }
        }
    }
    overlayCorners(content, regions, style);
    return content;
}

}  // namespace ui

File: java/src/main/java/uimd/FrameBufferView.java
package uimd;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Locale;
import java.util.Map;

public final class FrameBufferView extends Element
{
    private static final double DIM_ALPHA = 0.4;
    private static final Color DEFAULT_FOCUSED_COLOR = new Color("#facc15");
    private static final Color DEFAULT_EDIT_COLOR = new Color("#22c55e");

    private Map<String, List<List<FrameBufferCell>>> frames = new LinkedHashMap<>();
    private List<String> targetOrder = new ArrayList<>();
    private Map<String, FrameBufferRegion> regions = new LinkedHashMap<>();
    private int panRow;
    private int panCol;
    private boolean dimmed;
    private boolean panelFocused;
    private boolean panelEdit;

    public FrameBufferView(String name)
    {
        super(name);
    }

    public void clearFrames()
    {
        frames = new LinkedHashMap<>();
        targetOrder = new ArrayList<>();
        regions = new LinkedHashMap<>();
    }

    public void setFrames(
        Map<String, List<List<FrameBufferCell>>> value,
        List<String> order)
    {
        setFrames(value, order, Map.of());
    }

    public void setFrames(
        Map<String, List<List<FrameBufferCell>>> value,
        List<String> order,
        Map<String, FrameBufferRegion> targetRegions)
    {
        frames = copyFrames(value);
        targetOrder = order == null ? new ArrayList<>(frames.keySet()) : new ArrayList<>(order);
        regions = targetRegions == null
            ? new LinkedHashMap<>()
            : new LinkedHashMap<>(targetRegions);
    }

    public void setPan(int row, int col)
    {
        panRow = Math.max(0, row);
        panCol = Math.max(0, col);
    }

    public void setDimmed(boolean value)
    {
        dimmed = value;
    }

    public void setInteractionState(boolean focused, boolean edit)
    {
        panelFocused = focused;
        panelEdit = edit;
    }

    @Override
    public List<List<TerminalCell>> render(Size size, ElementRenderState state)
    {
        ElementRenderState effectiveState = state == null ? new ElementRenderState() : state;
        Style effective = effectiveStyle(effectiveState.focused(), effectiveState.editMode());
        int width = Math.max(1, size.width());
        int height = Math.max(1, size.height());
        List<List<TerminalCell>> content = RenderHelpers.renderPlainText(
            "",
            width,
            height,
            effective);
        Map<String, FrameBufferRegion> resolvedRegions = regions.isEmpty()
            ? defaultRegions(new Size(width, height))
            : regions;
        for (String target : targetOrder)
        {
            FrameBufferRegion region = resolvedRegions.get(target);
            List<List<FrameBufferCell>> rows = frames.get(target);
            if (region == null || rows == null)
            {
                continue;
            }
            renderTarget(content, rows, region, effective, width, height);
        }
        overlayCorners(content, resolvedRegions, effective);
        return content;
    }

    private void renderTarget(
        List<List<TerminalCell>> content,
        List<List<FrameBufferCell>> rows,
        FrameBufferRegion region,
        Style effective,
        int width,
        int height)
    {
        for (int localRow = 0; localRow < region.height(); ++localRow)
        {
            int outputRow = region.row() + localRow;
            int sourceRow = panRow + localRow;
            if (outputRow < 0 || outputRow >= height || sourceRow < 0 || sourceRow >= rows.size())
            {
                continue;
            }
            List<FrameBufferCell> sourceCells = rows.get(sourceRow);
            if (sourceCells == null)
            {
                continue;
            }
            for (int localCol = 0; localCol < region.width(); ++localCol)
            {
                int outputCol = region.col() + localCol;
                int sourceCol = panCol + localCol;
                if (outputCol < 0 || outputCol >= width
                    || sourceCol < 0 || sourceCol >= sourceCells.size())
                {
                    continue;
                }
                FrameBufferCell source = sourceCells.get(sourceCol);
                if (source == null)
                {
                    continue;
                }
                String foreground = dimmed ? dimHexColor(source.foreground()) : source.foreground();
                String background = dimmed ? dimHexColor(source.background()) : source.background();
                TerminalCell cell = new TerminalCell();
                cell.setText(RenderHelpers.safeTerminalCellText(source.text()));
                cell.setForeground(colorFromString(foreground, effective.color()));
                cell.setBackground(colorFromString(background, effective.background()));
                content.get(outputRow).set(outputCol, cell);
            }
        }
    }

    private Map<String, FrameBufferRegion> defaultRegions(Size size)
    {
        Map<String, FrameBufferRegion> result = new LinkedHashMap<>();
        List<Integer> widths = splitDimension(size.width(), Math.max(1, targetOrder.size()));
        int col = 0;
        for (int index = 0; index < targetOrder.size(); ++index)
        {
            int targetWidth = widths.get(index);
            result.put(
                targetOrder.get(index),
                new FrameBufferRegion(0, col, targetWidth, Math.max(1, size.height())));
            col += targetWidth;
        }
        return result;
    }

    private void overlayCorners(
        List<List<TerminalCell>> content,
        Map<String, FrameBufferRegion> resolvedRegions,
        Style effective)
    {
        if (!panelFocused && !panelEdit)
        {
            return;
        }
        Style stateStyle = panelEdit && editStyle() != null
            ? editStyle()
            : (panelFocused && focusStyle() != null ? focusStyle() : effective);
        Color color = stateStyle.color();
        if (color == null)
        {
            color = panelEdit ? DEFAULT_EDIT_COLOR : DEFAULT_FOCUSED_COLOR;
        }
        for (String target : targetOrder)
        {
            FrameBufferRegion region = resolvedRegions.get(target);
            if (region == null || region.width() < 2 || region.height() < 2)
            {
                continue;
            }
            putCorner(content, region.row(), region.col(), "┌", color);
            putCorner(content, region.row(), region.col() + 1, "─", color);
            putCorner(content, region.row(), region.col() + region.width() - 2, "─", color);
            putCorner(content, region.row(), region.col() + region.width() - 1, "┐", color);
            putCorner(content, region.row() + region.height() - 1, region.col(), "└", color);
            putCorner(content, region.row() + region.height() - 1, region.col() + 1, "─", color);
            putCorner(
                content,
                region.row() + region.height() - 1,
                region.col() + region.width() - 2,
                "─",
                color);
            putCorner(
                content,
                region.row() + region.height() - 1,
                region.col() + region.width() - 1,
                "┘",
                color);
        }
    }

    private static void putCorner(
        List<List<TerminalCell>> content,
        int row,
        int col,
        String text,
        Color color)
    {
        if (row < 0 || row >= content.size() || col < 0 || col >= content.get(row).size())
        {
            return;
        }
        TerminalCell cell = content.get(row).get(col);
        cell.setText(text);
        cell.setForeground(color);
    }

    private static List<Integer> splitDimension(int size, int parts)
    {
        int resolvedSize = Math.max(1, size);
        int resolvedParts = Math.max(1, Math.min(parts, resolvedSize));
        int base = resolvedSize / resolvedParts;
        int extra = resolvedSize % resolvedParts;
        List<Integer> result = new ArrayList<>(resolvedParts);
        for (int index = 0; index < resolvedParts; ++index)
        {
            result.add(base + (index < extra ? 1 : 0));
        }
        return result;
    }

    private static String dimHexColor(String value)
    {
        if (value == null || value.length() != 7 || value.charAt(0) != '#')
        {
            return value;
        }
        try
        {
            int red = dimChannel(value.substring(1, 3));
            int green = dimChannel(value.substring(3, 5));
            int blue = dimChannel(value.substring(5, 7));
            return String.format(Locale.ROOT, "#%02x%02x%02x", red, green, blue);
        }
        catch (NumberFormatException ignored)
        {
            return value;
        }
    }

    private static int dimChannel(String value)
    {
        return (int) Math.round((1.0 - DIM_ALPHA) * Integer.parseInt(value, 16));
    }

    private static Color colorFromString(String value, Color fallback)
    {
        if (value == null || value.isEmpty())
        {
            return fallback;
        }
        String resolved = value;
        if (value.length() == 4 && value.charAt(0) == '#')
        {
            resolved = "#"
                + value.charAt(1) + value.charAt(1)
                + value.charAt(2) + value.charAt(2)
                + value.charAt(3) + value.charAt(3);
        }
        try
        {
            return new Color(resolved);
        }
        catch (IllegalArgumentException ignored)
        {
            return fallback;
        }
    }

    private static Map<String, List<List<FrameBufferCell>>> copyFrames(
        Map<String, List<List<FrameBufferCell>>> value)
    {
        Map<String, List<List<FrameBufferCell>>> result = new LinkedHashMap<>();
        if (value == null)
        {
            return result;
        }
        for (Map.Entry<String, List<List<FrameBufferCell>>> entry : value.entrySet())
        {
            List<List<FrameBufferCell>> rows = new ArrayList<>();
            if (entry.getValue() != null)
            {
                for (List<FrameBufferCell> row : entry.getValue())
                {
                    rows.add(row == null ? new ArrayList<>() : new ArrayList<>(row));
                }
            }
            result.put(entry.getKey(), rows);
        }
        return result;
    }
}
